core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -162,72 +166,79 @@ final ClientTransport obtainActiveTransport() {
         transports.add(delayedTransport);
         delayedTransport.start(new BaseTransportListener(delayedTransport));
         activeTransport = delayedTransport;
-        scheduleConnection(delayedTransport);
+        startNewTransport(delayedTransport);
       }
       return activeTransport;
     }
   }
 
   @GuardedBy("lock")
-  private void scheduleConnection(final DelayedClientTransport delayedTransport) {
-    Preconditions.checkState(reconnectTask == null || reconnectTask.isDone(),
-        "previous reconnectTask is not done");
+  private void startNewTransport(DelayedClientTransport delayedTransport) {
+    Preconditions.checkState(reconnectTask == null, "Should have no reconnectTask scheduled");
 
-    if (firstAttempt) {
-      nextAddressIndex = 0;
+    if (nextAddressIndex == 0) {
+      connectingTimer.reset().start();
     }
-    final int currentAddressIndex = nextAddressIndex;
     List<SocketAddress> addrs = addressGroup.getAddresses();
-    final SocketAddress address = addrs.get(currentAddressIndex);
-    nextAddressIndex++;
+    final SocketAddress address = addrs.get(nextAddressIndex++);
     if (nextAddressIndex >= addrs.size()) {
       nextAddressIndex = 0;
     }
 
-    Runnable createTransportRunnable = new Runnable() {
+    ManagedClientTransport transport = transportFactory.newClientTransport(address, authority);
+    if (log.isLoggable(Level.FINE)) {
+      log.log(Level.FINE, "[{0}] Created {1} for {2}",
+          new Object[] {getLogId(), transport.getLogId(), address});
+    }
+    transports.add(transport);
+    transport.start(new TransportListener(transport, delayedTransport, address));
+  }
+
+  /**
+   * Only called after all addresses attempted and failed.
+   */
+  @GuardedBy("lock")
+  private void scheduleBackoff(final DelayedClientTransport delayedTransport) {
+    Preconditions.checkState(reconnectTask == null, "previous reconnectTask is not done");
+
+    if (reconnectPolicy == null) {
+      reconnectPolicy = backoffPolicyProvider.get();
+    }
+    long delayMillis =
+        reconnectPolicy.nextBackoffMillis() - connectingTimer.elapsed(TimeUnit.MILLISECONDS);
+    if (log.isLoggable(Level.FINE)) {
+      log.log(Level.FINE, "[{0}] Scheduling backoff for {1} ms",
+          new Object[]{getLogId(), delayMillis});
+    }
+    Runnable endOfCurrentBackoff = new Runnable() {
       @Override
       public void run() {
-        synchronized (lock) {
-          reconnectTask = null;
-          if (currentAddressIndex == 0) {
-            backoffWatch.reset().start();
+        try {
+          synchronized (lock) {
+            reconnectTask = null;
+            if (delayedTransport.hasPendingStreams()) {
+              startNewTransport(delayedTransport);
+            } else {
+              // Transition to IDLE (or already SHUTDOWN)
+              activeTransport = null;
+              delayedTransport.setTransportSupplier(new Supplier<ClientTransport>() {
```

Comment:
`setTransportSupplier()` and `shutdown()` don't need to be called under `lock`. Doesn't seem to cause a problem here, but it may be better to do it outside of `lock`.